mod/peer/gossip/state/state.go

```diff
@@ -52,34 +69,158 @@ type GossipServiceMediator interface {
 }
 
 //AddBlockHandler handles state update in gossip
-func AddBlockHandler(cid string, publisher api.BlockPublisher) {
-	//do nothing
+func AddBlockHandler(publisher api.BlockPublisher) {
+	publisher.AddWriteHandler(func(blockNum uint64, channelID, txID, namespace string, kvWrite *kvrwset.KVWrite) error {
+		if namespace != "lscc" {
+			return nil
+		}
+		return handleStateUpdate(kvWrite, channelID)
+	})
 }
 
 //NewGossipStateProviderExtension returns new GossipStateProvider Extension implementation
 func NewGossipStateProviderExtension(chainID string, mediator GossipServiceMediator) GossipStateProviderExtension {
-	return &gossipStateProviderExtension{}
+	return &gossipStateProviderExtension{chainID, mediator}
 }
 
 type gossipStateProviderExtension struct {
+	chainID  string
+	mediator GossipServiceMediator
 }
 
 func (s *gossipStateProviderExtension) HandleStateRequest(handle func(msg protoext.ReceivedMessage)) func(msg protoext.ReceivedMessage) {
-	return handle
+	return func(msg protoext.ReceivedMessage) {
+		if roles.IsEndorser() {
+			handle(msg)
+		}
+	}
 }
 
 func (s *gossipStateProviderExtension) AntiEntropy(handle func()) func() {
-	return handle
+	return func() {
+		if roles.IsCommitter() {
+			handle()
+		}
+	}
 }
 
 func (s *gossipStateProviderExtension) Predicate(handle func(peer discovery.NetworkMember) bool) func(peer discovery.NetworkMember) bool {
-	return handle
+	return func(peer discovery.NetworkMember) bool {
+		canPredicate := handle(peer)
+		if canPredicate {
+			if len(peer.Properties.Roles) == 0 || roles.HasEndorserRole(peer.Properties.Roles) {
+				logger.Debugf("Choosing [%s] since it's an endorser", peer.Endpoint)
+				return true
+			}
+			logger.Debugf("Not choosing [%s] since it's not an endorser", peer.Endpoint)
+			return false
+		}
+		return false
+	}
 }
 
 func (s *gossipStateProviderExtension) AddPayload(handle func(payload *proto.Payload, blockingMode bool) error) func(payload *proto.Payload, blockingMode bool) error {
-	return handle
+
+	return func(payload *proto.Payload, blockingMode bool) error {
+		if roles.IsCommitter() {
+			return handle(payload, blockingMode)
+		}
+		return nil
+	}
 }
 
 func (s *gossipStateProviderExtension) StoreBlock(handle func(block *common.Block, pvtData util.PvtDataCollections) error) func(block *common.Block, pvtData util.PvtDataCollections) error {
-	return handle
+
+	return func(block *common.Block, pvtData util.PvtDataCollections) error {
+		if roles.IsCommitter() {
+			fmt.Println("supposed to return error")
```

Comment:
Should use logger.

Reply:
Sorry, leftover from test run, just removed it.